ocaml/xapi/extauth_plugin_ADwinbind.ml

```diff
@@ -1440,33 +1301,65 @@ let build_dns_hostname_option ~config_params =
   | _ ->
       []
 
-let closest_kdc_of_domain domain =
-  match ClosestKdc.from_db domain with
-  | Some kdc when Result.is_ok (workgroup_from_server kdc) ->
-      kdc
-  | _ ->
-      (* Just pick the first valid KDC in the list *)
-      kdc_of_domain domain
+let domain_name_of_netbios netbios =
+  (*
+   * Query the domain name from netbios name with caching
+   * Check cache first, if not found, perform LDAP query and cache the result
+   *)
+  match Hashtbl.find_opt domain_netbios_name_map netbios with
+  | Some domain_name ->
+      debug "Cache hit for netbios '%s' -> domain '%s'" netbios domain_name ;
+      Ok domain_name
+  | None -> (
+      let {service_name; workgroup; _} = get_domain_info_from_db () in
+      match netbios = Option.value workgroup ~default:"" with
+      | true ->
+          Hashtbl.replace domain_netbios_name_map netbios service_name ;
```

Comment:
I can't find how will the keys in `domain_netbios_name_map` be removed.
Is it acceptable in terms of practical use cases that keeping its size non-decreasing until next XAPI restart?